### PR TITLE
UCP: tag matching offload eager support

### PR DESCRIPTION
**Pull request by alex-mikheev**
@yosefe please take a look

### CODE REVIEW

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/3676/ for details (Mellanox internal link).

File: src/ucp/tag/offload.c [hunk 1]
```diff
@@ -0,0 +1,251 @@
...
+                                            req->send.tag, imm_data,
+                                            pack_cb, req);
+    if (packed_len < 0) {
+        return packed_len;
+    }
+    return UCS_OK;
+}
+
+static UCS_F_ALWAYS_INLINE ucs_status_t
+ucp_do_tag_offload_zcopy(uct_pending_req_t *self, uint64_t imm_data,
```
Comment (yosefe) [inline src/ucp/tag/offload.c]: why need to define additional '_do' functions? they are used only once anyway

File: src/ucp/tag/tag_match.inl [hunk 2]
```diff
@@ -196,7 +197,12 @@ ucp_tag_unexp_desc_release(ucp_recv_desc_t *rdesc)
 {
     ucs_trace_req("release receive descriptor %p", rdesc);
     if (ucs_unlikely(rdesc->flags & UCP_RECV_DESC_FLAG_UCT_DESC)) {
-        uct_iface_release_desc(rdesc); /* uct desc is slowpath */
+        /* uct desc is slowpath */
+        if (rdesc->flags & UCP_RECV_DESC_FLAG_OFFLOAD) {
```
Comment (yosefe) [inline src/ucp/tag/tag_match.inl]: ucs_unlikely

File: src/ucp/tag/tag_recv.c [hunk 3]
```diff
@@ -123,6 +126,7 @@ ucp_tag_recv_request_init(ucp_request_t *req, ucp_worker_h worker, void* buffer,
     ucp_dt_generic_t *dt_gen;
     req->flags = UCP_REQUEST_FLAG_EXPECTED | UCP_REQUEST_FLAG_RECV | req_flags;
     req->recv.state.offset = 0;
+    req->recv.worker       = worker;
```
Comment (yosefe) [inline src/ucp/tag/tag_recv.c]: can use ucs_mpool_obj_owner() instead of adding field to recv request

File: src/ucp/tag/tag_recv.c [hunk 4]
```diff
@@ -202,6 +209,14 @@ ucp_tag_recv_common(ucp_worker_h worker, void *buffer, size_t count,
         req->recv.tag_mask = tag_mask;
         req->recv.cb       = cb;
         ucp_tag_exp_push(&context->tm, queue, req);
+
+        if (!(req->flags & UCP_REQUEST_FLAG_MATCHED)) {
+            /* If offload supported, post this tag to transport as well.
+             * Check for match flag, because it may be a part of already
+             * matched message in AM protocol (thus we do not need to post
```
Comment (yosefe) [inline src/ucp/tag/tag_recv.c]: how can this be ? if offload is enabled, the sender should not send multi-fragment message at all.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1738/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/3683/ for details (Mellanox internal link).

**Comment by yosefe** [comment]
bot:bgate:retest

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1743/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1747/ for details.

**Comment by yosefe** [comment]
bot:bgate:retest

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1748/ for details.

**Pull request merged**